src/v/net/probes.cc

```diff
@@ -282,4 +299,187 @@ std::ostream& operator<<(std::ostream& o, const client_probe& p) {
+  std::string_view service,
+  std::string_view listener_name,
+  ss::tls::reload_callback cb) {
+    auto probe = ss::make_lw_shared<net::tls_certificate_probe>();
+    auto wrap_cb = [probe, cb = std::move(cb)](
+                     const std::unordered_set<ss::sstring>& updated,
+                     const ss::tls::certificate_credentials& creds,
+                     const std::exception_ptr& eptr) {
+        if (cb) {
+            cb(updated, eptr);
+        }
+        probe->loaded(creds, eptr);
+    };
+
+    ss::shared_ptr<T> cred;
+    if constexpr (std::is_same<T, ss::tls::server_credentials>::value) {
+        cred = co_await builder.build_reloadable_server_credentials(wrap_cb);
+    } else {
+        cred = co_await builder.build_reloadable_certificate_credentials(
+          wrap_cb);
+    }
+
+    probe->setup_metrics(service, listener_name);
+    probe->loaded(*cred, nullptr);
+    co_return cred;
+}
+
+template ss::future<ss::shared_ptr<ss::tls::server_credentials>>
+build_reloadable_credentials_with_probe(
+  ss::tls::credentials_builder builder,
+  std::string_view service,
+  std::string_view listener_name,
+  ss::tls::reload_callback cb);
+
+template ss::future<ss::shared_ptr<ss::tls::certificate_credentials>>
+build_reloadable_credentials_with_probe(
+  ss::tls::credentials_builder builder,
+  std::string_view service,
+  std::string_view listener_name,
+  ss::tls::reload_callback cb);
+
+void tls_certificate_probe::loaded(
+  const ss::tls::certificate_credentials& creds, std::exception_ptr ex) {
+    _load_time = clock_type::now();
+
+    if (ex) {
+        reset();
+        return;
+    }
+
+    auto to_tls_serial = [](bytes_view b) {
+        using T = tls_serial_number::type;
+        T result = 0;
+        const auto end = std::min(b.size(), sizeof(T));
+        std::memcpy(&result, b.data(), end);
+        return tls_serial_number{result};
+    };
+
+    _cert_loaded = true;
+
+    auto certs_info = creds.get_cert_info();
+    auto ts_info = creds.get_trust_list_info();
+
+    if (!certs_info.has_value() || !ts_info.has_value()) {
+        reset();
+        return;
+    }
+
+    for (auto& info : certs_info.value()) {
+        auto exp = clock_type::from_time_t(info.expiry);
+        auto srl = to_tls_serial(info.serial);
+        if (exp < _cert_expiry_time) {
+            _cert_expiry_time = exp;
+            _cert_serial = srl;
+        }
+    }
+
+    for (auto& info : ts_info.value()) {
+        auto exp = clock_type::from_time_t(info.expiry);
+        auto srl = to_tls_serial(info.serial);
+        if (exp < _ca_expiry_time) {
+            _ca_expiry_time = exp;
+            _ca_serial = srl;
+        }
+    }
+}
+
+void tls_certificate_probe::setup_metrics(
+  std::string_view area, std::string_view detail) {
+    if (ss::this_shard_id() != 0) {
+        return;
+    }
+
+    namespace sm = ss::metrics;
+    const auto area_label = sm::label("area");
+    const auto detail_label = sm::label("detail");
```

Comment:
It's not obvious to me what these names mean. service and listener name are clearer.

For the public metrics there's a wrapper for labels `make_namespace_label` to prevent collisions in prometheus for common labels such as "service", the goal was to use it for all labels on the public metrics.

Reply:
Ahh, I see, when it comes to client certs service and listener aren't so appropriate.

Reply:
Yeah, adding client credentials makes naming much harder. This way seems slightly less worse than alternatives (e.g. populating a subset of a larger group of labels conditional on type OR keeping service/listener and living with weird semantics). Very much open to suggestion.